### PR TITLE
Support restart and stop

### PR DESCRIPTION
**Pull request by testforstephen**

### CODE REVIEW

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/DebugException.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+* http://www.eclipse.org/legal/epl-v10.html
+*
+* Contributors:
+*     Microsoft Corporation - initial API and implementation
+*******************************************************************************/
+
+package org.eclipse.jdt.ls.debug;
+
+public class DebugException extends Exception {
+    private static final long serialVersionUID = 1L;
```
Comment (akaroml) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/DebugException.java:15]: What's this for?

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/DebugException.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+    public DebugException() {
```
Comment (akaroml) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/DebugException.java:17]: Maybe we can remove this. An exception without any description or cause should not be used.

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java [hunk 3]
```diff
@@ -246,64 +249,48 @@ public DebugAdapter(Consumer<Events.DebugEvent> consumer) {
...
-        }
-
-        if (arguments.sourcePath == null || arguments.sourcePath.length == 0) {
-            this.sourcePath = new String[] { cwd };
-        } else {
-            this.sourcePath = new String[arguments.sourcePath.length];
-            System.arraycopy(arguments.sourcePath, 0, this.sourcePath, 0, arguments.sourcePath.length);
+            this.launchDebugSession(arguments);
+        } catch (DebugException e) {
+            return new Responses.ErrorResponseBody(new Types.Message(3001, "Cannot launch debuggee vm: " + e.getMessage(), null));
```
Comment (akaroml) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java]: Still, don't understand the code 3001. Why do we need to specify this number?

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java [hunk 4]
```diff
@@ -246,64 +249,48 @@ public DebugAdapter(Consumer<Events.DebugEvent> consumer) {
...
     }
 
+    private Responses.ResponseBody restart(Requests.RestartArguments arguments) {
+        // Shutdown the old debug session.
+        this.shutdownDebugSession(true);
+        // Launch new debug session.
+        try {
+            this.launchDebugSession(this.launchArguments);
+        } catch (DebugException e) {
+            return new Responses.ErrorResponseBody(new Types.Message(3001, "Cannot restart debuggee vm: " + e.getMessage(), null));
```
Comment (akaroml) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java]: Same here.

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java [hunk 5]
```diff
@@ -466,10 +448,14 @@ private void handleEvent(DebugEvent debugEvent) {
             // do nothing.
         } else if (event instanceof VMDeathEvent || event instanceof VMDisconnectEvent) {
```
Comment (akaroml) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java]: VMDeathEvent and VMDisconnectEvent have different meanings and indicate different stages of termination. Please handle them in different code branch.

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/Events.java [hunk 6]
```diff
@@ -55,6 +55,35 @@ public StoppedEvent(String reason, Types.Source src, int ln, int col, String tex
...
     }
 
+    public static class ContinuedEvent extends DebugEvent {
+        public long threadId;
+        public boolean allThreadsContinued;
+
+        /**
+         * Constructor.
+         */
+        public ContinuedEvent(long id) {
```
Comment (akaroml) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/Events.java]: name the parameter threadId

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/Events.java [hunk 7]
```diff
@@ -55,6 +55,35 @@ public StoppedEvent(String reason, Types.Source src, int ln, int col, String tex
...
+            super("continued");
+            this.threadId = id;
+        }
+
+        /**
+         * Constructor.
+         */
+        public ContinuedEvent(long id, boolean allThreadsContinued) {
```
Comment (akaroml) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/Events.java]: threadId

**Review by akaroml — CHANGES_REQUESTED** [review]

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/DebugException.java (hunk 1 above)
Comment (testforstephen) [reply org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/DebugException.java:15]: DebugException extends Serializable, and Serializable should declare a field `serialVersionUID`.
https://stackoverflow.com/questions/285793/what-is-a-serialversionuid-and-why-should-i-use-it

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/DebugException.java (hunk 2 above)
Comment (testforstephen) [reply org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/DebugException.java:17]: DebugException extends Exception. And from javadoc of Exception's default constructor `java.lang.Exception.Exception()`, it said `
Constructs a new exception with null as its detail message. The cause is not initialized, and may subsequently be initialized by a call to initCause.
`

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java (hunk 3 above)
Comment (testforstephen) [reply org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java]: Accord to debugProtocol, in ErrorResponse body, we could set an optional structure Message. And the Message has a required id field. The code 3001 just reference to OpenDebugAD7 error code.

Comment (testforstephen) [reply org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java]: https://github.com/Microsoft/vscode-debugadapter-node/blob/master/protocol/src/debugProtocol.ts

**Review by akaroml — APPROVED** [review]
Nice work. Thanks.

**Pull request merged**